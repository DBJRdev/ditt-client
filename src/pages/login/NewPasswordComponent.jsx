import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import {
  Alert,
  Button,
  Card,
  CardBody,
  TextField,
} from '@react-ui-org/react-ui';
import {
  Icon,
  LoadingIcon,
} from '../../components/Icon';
import {
  SET_NEW_PASSWORD_SUCCESS,
  SET_NEW_PASSWORD_FAILURE,
} from '../../resources/auth/actionTypes';
import routes from '../../routes';
import LogoImage from './images/logo.svg';
import styles from './NewPassword.scss';

class NewPasswordComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: null,
      isSubmitted: false,
      newPassword: null,
      newPasswordRepeat: null,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.onSubmitHandler = this.onSubmitHandler.bind(this);
    this.newPasswordHandler = this.newPasswordHandler.bind(this);
  }

  onChangeHandler(e) {
    this.setState({
      [e.target.id]: e.target.value,
    });
  }

  onSubmitHandler(e) {
    e.preventDefault();

    this.newPasswordHandler(e);
  }

  newPasswordHandler(e) {
    if (this.state.newPassword !== this.state.newPasswordRepeat) {
      e.preventDefault();
      this.setState({ error: this.props.t('login:validation.passwordsNotSame') });

      return;
    }

    if (
      this.state.newPassword
      && (this.state.newPassword.length < 8 || this.state.newPassword.length > 64)
    ) {
      e.preventDefault();
      this.setState({
        error: this.props.t(
          'login:validation.passwordInvalidMinLength',
          {
            max: 64,
            min: 8,
          },
        ),
      });

      return;
    }

    this.setState({ error: null });

    this.props.setNewPassword({
      newPlainPassword: this.state.newPassword,
      resetPasswordToken: this.props.match.params.resetPasswordToken,
    }).then((response) => {
      if (response.type === SET_NEW_PASSWORD_SUCCESS) {
        this.setState({ isSubmitted: true });
      } else if (response.type === SET_NEW_PASSWORD_FAILURE) {
        this.setState({ error: response.payload.response.detail });
      }
    });
  }

  render() {
    const {
      isPosting,
      t,
    } = this.props;
    const {
      error,
      newPassword,
      newPasswordRepeat,
      isSubmitted,
    } = this.state;

    const layout = (children) => (
      <div className={styles.container}>
        <LogoImage
          alt={t('layout:title')}
          className={styles.logo}
          height={141}
          width={302}
        />
        {children}
      </div>
    );

    if (isSubmitted) {
      return layout((
        <p className={styles.message}>
          {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
          {t('login:text.passwordSet')}
          {' '}
          <Link to={routes.login}>{t('login:action.login')}</Link>
        </p>
      ));
    }

    return layout((
      <div className={styles.box}>
        <div className={styles.title}>
          {t('layout:title')}
        </div>
        {error && (
          <div className="mb-5">
            <Alert
              icon={<Icon icon="error" />}
              type="error"
            >
              <strong>
                {t('general:text.error')}
                {': '}
              </strong>
              {error}
            </Alert>
          </div>
        )}
        <Card variant="bordered">
          <CardBody>
            <form
              onSubmit={this.onSubmitHandler}
            >
              <div className="mb-3">
                <TextField
                  autoComplete="new-password"
                  changeHandler={this.onChangeHandler}
                  fullWidth
                  id="newPassword"
                  label={t('login:element.newPassword')}
                  type="password"
                  required
                  value={newPassword ?? ''}
                />
                <TextField
                  autoComplete="new-password"
                  changeHandler={this.onChangeHandler}
                  fullWidth
                  id="newPasswordRepeat"
                  label={t('login:element.repeatNewPassword')}
                  type="password"
                  required
                  value={newPasswordRepeat ?? ''}
                />
              </div>
              <Button
                block
                label={t('login:action.changePassword')}
                loadingIcon={isPosting && <LoadingIcon />}
                type="submit"
              />
            </form>
            <div className={styles.footer}>
              <Link to={routes.login}>
                {t('login:action.login')}
              </Link>
            </div>
          </CardBody>
        </Card>
      </div>
    ));
  }
}

NewPasswordComponent.propTypes = {
  isPosting: PropTypes.bool.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      resetPasswordToken: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  setNewPassword: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired,
};

export default withTranslation()(NewPasswordComponent);
