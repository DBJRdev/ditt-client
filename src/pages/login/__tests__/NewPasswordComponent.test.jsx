import React from 'react';
import { shallow } from 'enzyme';
import NewPasswordComponent from '../NewPasswordComponent';

describe('rendering', () => {
  it('renders correctly', () => {
    const props = {
      isPosting: false,
      isPostingFailure: false,
      match: {
        params: {
          resetPasswordToken: 'token',
        },
      },
      newPassword: () => {},
    };
    const tree = shallow(<NewPasswordComponent {...props} />);

    expect(tree).toMatchSnapshot();
  });
});
