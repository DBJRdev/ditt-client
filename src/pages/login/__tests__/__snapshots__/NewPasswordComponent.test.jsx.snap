// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<div
  className="container"
>
  <img
    alt="layout:title"
    className="logo"
    height={141}
    src={Object {}}
    width={302}
  />
  <div
    className="NewPassword__root__3WHH0t8q"
  >
    <div
      className="NewPassword__title__KAzfLVH6"
    >
      layout:title
    </div>
    <div
      className="NewPassword__box__i2dyGe6h"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="TextField__root__3yDmdJNi
        TextField__isRootFullWidth__1hLzNevO
        
        TextField__rootLayoutVertical__3YCcUzUr
        TextField__isRootRequired__FUf0CdvL
        TextField__rootSizeMedium__3o17OJby
        
        
        TextField__rootVariantOutline__1RuTCzuC"
            htmlFor="newPasswordInput"
            id="newPasswordInput__label"
            style={Object {}}
          >
            <div
              className="TextField__label__2bJzQO6p"
              id="newPasswordInput__labelText"
            >
              login:element.newPassword
            </div>
            <div
              className="TextField__field__1Nm2afhb"
            >
              <div
                className="TextField__inputContainer__3NDuxsmR"
              >
                <input
                  autoComplete="new-password"
                  className="TextField__input__dZrOwzNX"
                  disabled={false}
                  id="newPasswordInput"
                  onChange={[Function]}
                  placeholder={null}
                  required={true}
                  size={null}
                  type="password"
                />
              </div>
            </div>
          </label>
          <label
            className="TextField__root__3yDmdJNi
        TextField__isRootFullWidth__1hLzNevO
        
        TextField__rootLayoutVertical__3YCcUzUr
        TextField__isRootRequired__FUf0CdvL
        TextField__rootSizeMedium__3o17OJby
        
        
        TextField__rootVariantOutline__1RuTCzuC"
            htmlFor="newPasswordRepeatInput"
            id="newPasswordRepeatInput__label"
            style={Object {}}
          >
            <div
              className="TextField__label__2bJzQO6p"
              id="newPasswordRepeatInput__labelText"
            >
              login:element.repeatNewPassword
            </div>
            <div
              className="TextField__field__1Nm2afhb"
            >
              <div
                className="TextField__inputContainer__3NDuxsmR"
              >
                <input
                  autoComplete="new-password"
                  className="TextField__input__dZrOwzNX"
                  disabled={false}
                  id="newPasswordRepeatInput"
                  onChange={[Function]}
                  placeholder={null}
                  required={true}
                  size={null}
                  type="password"
                />
              </div>
            </div>
          </label>
        </div>
        <button
          className="Button__root__3zrxRSxl
        Button__priorityDefault__hcWO8Peq
        Button__sizeMedium__168V18-G
        Button__variantPrimary__30PKnizS
        Button__isRootBlock__dk3xheVK"
          disabled={false}
          onClick={null}
          title={null}
          type="submit"
        >
          <span
            className="Button__label__3YN6eZ6q"
          >
            login:action.changePassword
          </span>
        </button>
      </form>
      <div
        className="NewPassword__footer__1RBGwF6G"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          login:action.login
        </a>
      </div>
    </div>
  </div>
</div>
`;
